service/frontend/versionChecker.go

```diff
@@ -0,0 +1,168 @@
+	if vc.config.EnableServerVersionCheck() {
+		vc.startOnce.Do(func() {
+			go vc.versionCheckLoop()
+		})
+	}
+}
+
+func (vc *VersionChecker) Stop() {
+	if vc.config.EnableServerVersionCheck() {
+		vc.stopOnce.Do(func() {
+			close(vc.shutdownChan)
+		})
+	}
+}
+
+func (vc *VersionChecker) versionCheckLoop() {
+	timer := time.NewTicker(VersionCheckInterval)
+	defer timer.Stop()
+	vc.performVersionCheck()
+	for {
+		select {
+		case <-vc.shutdownChan:
+			return
+		case <-timer.C:
+			vc.performVersionCheck()
+		}
+	}
+}
+
+func (vc *VersionChecker) performVersionCheck() {
+	sw := vc.metricsScope.StartTimer(metrics.VersionCheckLatency)
+	defer sw.Stop()
+	clusterMetadataManager := vc.GetClusterMetadataManager()
+	metadata, err := clusterMetadataManager.GetClusterMetadata()
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+
+	if !isUpdateNeeded(metadata) {
+		return
+	}
+
+	req, err := vc.createVersionCheckRequest(metadata)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+	resp, err := vc.getVersionInfo(req)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+	err = vc.saveVersionInfo(resp)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+	vc.metricsScope.IncCounter(metrics.VersionCheckSuccessCount)
+}
+
+func isUpdateNeeded(metadata *persistence.GetClusterMetadataResponse) bool {
+	return metadata.VersionInfo == nil || (metadata.VersionInfo.LastUpdateTime != nil &&
+		metadata.VersionInfo.LastUpdateTime.Before(time.Now().Add(-time.Hour)))
+}
+
+func (vc *VersionChecker) createVersionCheckRequest(metadata *persistence.GetClusterMetadataResponse) (*check.VersionCheckRequest, error) {
+	return &check.VersionCheckRequest{
+		Product:   headers.ClientNameServer,
+		Version:   headers.ServerVersion,
+		Arch:      runtime.GOARCH,
+		OS:        runtime.GOOS,
+		DB:        vc.GetClusterMetadataManager().GetName(),
+		ClusterID: metadata.ClusterId,
+		Timestamp: time.Now().UnixNano(),
+	}, nil
+}
+
+func (vc *VersionChecker) getVersionInfo(req *check.VersionCheckRequest) (*check.VersionCheckResponse, error) {
+	return check.NewCaller().Call(req)
+}
+
+func (vc *VersionChecker) saveVersionInfo(resp *check.VersionCheckResponse) error {
+	clusterMetadataManager := vc.GetClusterMetadataManager()
+	metadata, err := clusterMetadataManager.GetClusterMetadata()
+	if err != nil {
+		return err
+	}
+	metadata.VersionInfo = toVersionInfo(resp)
+	saved, err := clusterMetadataManager.SaveClusterMetadata(&persistence.SaveClusterMetadataRequest{
+		ClusterMetadata: metadata.ClusterMetadata, Version: metadata.Version})
+	if err != nil {
+		return err
+	}
+	if !saved {
+		return serviceerror.NewInternal("version info update hasn't been applied")
+	}
+	return nil
+}
+
+func toVersionInfo(resp *check.VersionCheckResponse) *persistenceblobs.VersionInfo {
+	currentTime := time.Now()
```

Comment:
I would suggest to always add `.UTC()` to `time.Now()`. It probably doesn't matter here but it is a good habit.

Reply:
Done